File: sdk/src/main/java/com/libreshockwave/chunks/KeyTableChunk.java
package com.libreshockwave.chunks;

import com.libreshockwave.DirectorFile;
import com.libreshockwave.format.ChunkType;
import com.libreshockwave.id.ChunkId;
import com.libreshockwave.io.BinaryReader;

import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Key table chunk (KEY*).
 * Maps resource IDs to their associated chunks.
 * Provides bidirectional lookups:
 * - Owner to chunks: getEntriesForOwner(castId) returns all chunks owned by a cast member
 * - Chunk to owner: getOwnerCastId(sectionId) returns the cast member that owns a chunk
 */
public record KeyTableChunk(
    DirectorFile file,
    ChunkId id,
    List<KeyTableEntry> entries,
    Map<ChunkId, List<KeyTableEntry>> entriesByOwner,
    Map<ChunkId, ChunkId> ownerBySectionId
) implements Chunk {

    @Override
    public ChunkType type() {
        return ChunkType.KEYp;
    }

    public record KeyTableEntry(
        ChunkId sectionId,
        ChunkId castId,
        int fourcc
    ) {
        public String fourccString() {
            return BinaryReader.fourCCToString(fourcc);
        }
    }

    public List<KeyTableEntry> getEntriesForOwner(ChunkId ownerId) {
        return entriesByOwner.getOrDefault(ownerId, List.of());
    }

    public KeyTableEntry findEntry(ChunkId ownerId, int fourcc) {
        List<KeyTableEntry> ownerEntries = entriesByOwner.get(ownerId);
        if (ownerEntries == null) return null;

        for (KeyTableEntry entry : ownerEntries) {
            if (entry.fourcc == fourcc) {
                return entry;
            }
        }
        return null;
    }

    /**
     * Get the cast member ID that owns a chunk by its section ID.
     * @param sectionId The chunk's section ID (resource ID)
     * @return The owning cast member's ID, or null if not found
     */
    public ChunkId getOwnerCastId(ChunkId sectionId) {
        return ownerBySectionId.get(sectionId);
    }

    /**
     * Get the KeyTableEntry for a chunk by its section ID.
     * @param sectionId The chunk's section ID (resource ID)
     * @return The KeyTableEntry, or null if not found
     */
    public KeyTableEntry getEntryBySectionId(ChunkId sectionId) {
        ChunkId ownerId = ownerBySectionId.get(sectionId);
        if (ownerId == null) {
            return null;
        }
        for (KeyTableEntry entry : entries) {
            if (entry.sectionId.equals(sectionId)) {
                return entry;
            }
        }
        return null;
    }

    public static KeyTableChunk read(DirectorFile vm, BinaryReader reader, ChunkId id, int version) {
        ByteOrder originalOrder = reader.getOrder();
        reader.setOrder(ByteOrder.LITTLE_ENDIAN);

        int headerSize = reader.readI16() & 0xFFFF;
        int entrySize = reader.readI16() & 0xFFFF;
        int totalCount = reader.readI32();
        int usedCount = reader.readI32();

        List<KeyTableEntry> entries = new ArrayList<>();
        Map<ChunkId, List<KeyTableEntry>> byOwner = new HashMap<>();
        Map<ChunkId, ChunkId> ownerBySection = new HashMap<>();

        for (int i = 0; i < usedCount; i++) {
            // Key table entries may have negative sentinel values; clamp to 0
            ChunkId sectionId = new ChunkId(Math.max(0, reader.readI32()));
            ChunkId castId = new ChunkId(Math.max(0, reader.readI32()));
            // FourCC in key table is stored as a u32 in the file's byte order (little-endian).
            // In LE files, tags are byte-swapped (e.g., "ALFA" stored as "AFLA" bytes).
            // readI32() with LE byte order produces the correct integer for fourCCToString.
            int fourcc = reader.readI32();

            KeyTableEntry entry = new KeyTableEntry(sectionId, castId, fourcc);
            entries.add(entry);

            byOwner.computeIfAbsent(castId, k -> new ArrayList<>()).add(entry);
            ownerBySection.put(sectionId, castId);
        }

        reader.setOrder(originalOrder);

        return new KeyTableChunk(vm, id, entries, byOwner, ownerBySection);
    }
}

File: cpp/src/chunks/KeyTableChunk.cpp
#include "libreshockwave/chunks/KeyTableChunk.hpp"

#include <algorithm>

#include "libreshockwave/io/BinaryReader.hpp"

namespace libreshockwave::chunks {

std::string KeyTableChunk::KeyTableEntry::fourccString() const {
    return io::BinaryReader::fourCCToString(fourcc);
}

KeyTableChunk::KeyTableChunk(const DirectorFile* file, id::ChunkId id, std::vector<KeyTableEntry> entries)
    : file_(file), id_(id), entries_(std::move(entries)) {
    indexEntries();
}

const DirectorFile* KeyTableChunk::file() const { return file_; }
format::ChunkType KeyTableChunk::type() const { return format::ChunkType::KEYp; }
id::ChunkId KeyTableChunk::id() const { return id_; }
const std::vector<KeyTableChunk::KeyTableEntry>& KeyTableChunk::entries() const { return entries_; }

std::vector<KeyTableChunk::KeyTableEntry> KeyTableChunk::getEntriesForOwner(id::ChunkId ownerId) const {
    if (const auto found = entriesByOwner_.find(ownerId.value()); found != entriesByOwner_.end()) {
        return found->second;
    }
    return {};
}

std::optional<KeyTableChunk::KeyTableEntry> KeyTableChunk::findEntry(id::ChunkId ownerId, std::uint32_t fourcc) const {
    const auto entries = getEntriesForOwner(ownerId);
    for (const auto& entry : entries) {
        if (entry.fourcc == fourcc) {
            return entry;
        }
    }
    return std::nullopt;
}

std::optional<id::ChunkId> KeyTableChunk::getOwnerCastId(id::ChunkId sectionId) const {
    if (const auto found = ownerBySectionId_.find(sectionId.value()); found != ownerBySectionId_.end()) {
        return id::ChunkId(found->second);
    }
    return std::nullopt;
}

std::optional<KeyTableChunk::KeyTableEntry> KeyTableChunk::getEntryBySectionId(id::ChunkId sectionId) const {
    if (!getOwnerCastId(sectionId).has_value()) {
        return std::nullopt;
    }
    for (const auto& entry : entries_) {
        if (entry.sectionId.value() == sectionId.value()) {
            return entry;
        }
    }
    return std::nullopt;
}

KeyTableChunk KeyTableChunk::read(const DirectorFile* file, io::BinaryReader& reader, id::ChunkId id, int version) {
    (void)version;
    (void)reader.readU16();
    (void)reader.readU16();
    (void)reader.readI32();
    const int usedCount = reader.readI32();

    std::vector<KeyTableEntry> entries;
    for (int index = 0; index < usedCount; ++index) {
        auto sectionId = id::ChunkId(sanitizeChunkId(reader.readI32()));
        auto castId = id::ChunkId(sanitizeChunkId(reader.readI32()));
        const auto fourcc = reader.readU32();
        entries.push_back(KeyTableEntry{sectionId, castId, fourcc});
    }
    return KeyTableChunk(file, id, std::move(entries));
}

int KeyTableChunk::sanitizeChunkId(int value) {
    return std::max(0, value);
}

void KeyTableChunk::indexEntries() {
    entriesByOwner_.clear();
    ownerBySectionId_.clear();
    for (const auto& entry : entries_) {
        entriesByOwner_[entry.castId.value()].push_back(entry);
        ownerBySectionId_[entry.sectionId.value()] = entry.castId.value();
    }
}

} // namespace libreshockwave::chunks
